packages/map-toolkit/src/deckgl/base-map/map-mode-provider/index.tsx

```diff
@@ -0,0 +1,272 @@
+ *   });
+ *
+ *   return (
+ *     <button onClick={() => requestModeChange('drawing', 'drawing-tool')}>
+ *       Start Drawing
+ *     </button>
+ *   );
+ * }
+ * ```
+ */
+export function MapModeProvider({
+  children,
+  defaultMode = DEFAULT_MODE,
+}: MapModeProviderProps) {
+  const [mode, setMode] = useState(defaultMode);
+  // Store mode-to-owner mappings (persists throughout session, no re-renders needed)
+  const modeOwnersRef = useRef<Map<string, string>>(new Map());
+  // Store pending authorization requests by authId (no re-renders needed)
+  const pendingRequestsRef = useRef<Map<string, PendingRequest>>(new Map());
+
+  const emitChanged = useEmit<ModeChangedEvent>(MapModeEvents.changed);
+  const emitAuthorization = useEmit<ModeChangeAuthorizationEvent>(
+    MapModeEvents.changeAuthorization,
+  );
+  const emitRequest = useEmit<ModeChangeRequestEvent>(
+    MapModeEvents.changeRequest,
+  );
+
+  // Centralized listener for mode change requests
+  useOn<ModeChangeRequestEvent>(
+    MapModeEvents.changeRequest,
+    (event: ModeChangeRequestEvent) => {
+      const { desiredMode, owner: requestOwner } = event.payload;
+
+      // If current mode is same as desired, no change needed
+      if (desiredMode === mode) {
+        return;
+      }
+
+      // Get current mode's owner from the mapping
+      const currentModeOwner = modeOwnersRef.current.get(mode);
+      // Get desired mode's owner from the mapping
+      const desiredModeOwner = modeOwnersRef.current.get(desiredMode);
+
+      // Auto-accept if:
+      // 1. Desired mode is 'default' AND requesting owner is the current mode owner
+      // 2. Requesting owner is same as current mode owner, OR
+      // 3. No current or desired mode owner, OR
+      // 4. In default mode and the requester is the owner of the desired mode
+      if (
+        (desiredMode === defaultMode && requestOwner === currentModeOwner) ||
+        requestOwner === currentModeOwner ||
+        !(currentModeOwner || desiredModeOwner) ||
+        (mode === defaultMode && requestOwner === desiredModeOwner)
+      ) {
+        setMode(desiredMode);
+        emitChanged({
+          previousMode: mode,
+          currentMode: desiredMode,
+        });
+
+        // Store the desired mode's owner unless it's default (which stays ownerless),
+        // and there's not another mode owner already
+        if (desiredMode !== defaultMode && !desiredModeOwner) {
+          modeOwnersRef.current.set(desiredMode, requestOwner);
+        }
+
+        // Clear any pending authorization requests since mode changed successfully
+        pendingRequestsRef.current.clear();
+        return;
+      }
+
+      // Otherwise, send authorization request
+      // Generate authId and store the request for later lookup
+      const authId = uuid();
+
+      // Clear any existing pending requests and add the new one (only one pending request at a time)
```

Comment:
You've got a mixed strategy going on. Right now this clear is only allowing a single request at a time. But you're using a map which implies support for multiple requests

As I intended it, you'd want to allow multiple requests, but only one per owner, which means storing it by owner id, not auth id. The auth id is just used to make sure that the decision aligns with an active pending request

Though I think you've touched on an interesting edge case as well, where we'd likely want to clear the map of requests when a decision has been made.

Reply:
ah - good catch. I'll update

Reply:
should pending requests be canceled if the owner changes the mode to a different one before dealing with a pending request?

Reply:
answered in chat: one per requester, don't cancel pending requests if owner changes modes unless they change to default and the request is for default (same mode)